time-it-builder/src/lib.rs

```diff
@@ -37,37 +77,123 @@ impl TimeItBuilder {
+                    while !(shutdown.load(Ordering::Relaxed) && receiver.is_empty()) {
+                        match receiver.recv() {
+                            Ok(event) => {
+                                (lazy_config.action)(
+                                    event.tag.as_ref().unwrap_or(&default_tag),
+                                    event.duration,
+                                );
+                            }
+                            Err(_) => {
+                                break;
+                            }
+                        }
+                    }
+                });
+                lazy_config.sender = Some(sender);
+                // Keep the config in the original struct because we want to get mutable reference to it later for cleaning up
+                config.receiver_handle = Some(handle);
+                lazy_config
+            }
+            TimeItMode::Sync => lazy_config,
+        }
+    })
+}
+
+pub fn run(tag: Option<&str>, duration: Duration) {
+    let config = get_config();
+    match config.mode {
+        TimeItMode::Async => {
+            emit(tag, duration);
+        }
+        TimeItMode::Sync => {
+            action(tag, duration);
+        }
+    }
+}
+
+pub fn wait_for_finish() {
+    let mut config = TIMEIT_CONFIG.write();
+    config.shutdown.store(true, Ordering::Relaxed);
+    config.wait_for_finish()
 }
 
 pub fn action(tag: Option<&str>, duration: Duration) {
     let config = get_config();
     (config.action)(tag.unwrap_or(&config.default_tag), duration);
 }
 
+pub fn emit(tag: Option<&str>, duration: Duration) {
+    let config = get_config();
+    let sender = config.sender.as_ref().unwrap();
+    sender
+        .send(TimeItEvent {
+            tag: tag.map(|s| s.to_string()),
+            duration,
+        })
+        .unwrap();
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
 
     #[test]
     fn it_works() {
-        TimeItBuilder::new().time_it(|tag, duration| {
-            let millis = duration.as_millis();
-            println!("Time Elapsed during {tag}: {}ms", millis)
-        });
+        TimeItBuilder::new();
+    }
+
+    #[test]
+    fn it_works_async() {
+        TimeItBuilder::new()
+            .time_it(|tag, duration| {
+                let millis = duration.as_millis();
+                println!("Time Elapsed during {tag}: {}ms", millis)
+            })
+            .mode(TimeItMode::Async);
+
+        let config = self::get_config();
+        let action = config.action.clone();
+        action("Custom Message", Duration::from_millis(100));
+    }
+
+    #[test]
+    fn it_works_sync() {
+        TimeItBuilder::new()
+            .time_it(|tag, duration| {
+                let millis = duration.as_millis();
+                println!("Time Elapsed during {tag}: {}ms", millis)
+            })
+            .mode(TimeItMode::Sync);
 
         let config = self::get_config();
         let action = config.action.clone();
```

Comment:
+1 on calling 1

Reply:
This is just testing getting the action out of the config, this is the field in the struct, not the method that uses the global static.